client/resoucemanager_client.go

```diff
@@ -0,0 +1,265 @@
+}
+
+func (c *client) putResourceGroup(ctx context.Context, resourceGroupName string, settings *rmpb.GroupSettings, typ int) (str string, err error) {
+	group := &rmpb.ResourceGroup{
+		Name:     resourceGroupName,
+		Settings: settings,
+	}
+	req := &rmpb.PutResourceGroupRequest{
+		Group: group,
+	}
+	var resp *rmpb.PutResourceGroupResponse
+	if typ == 0 {
+		resp, err = c.resouceManagerClient().AddResourceGroup(ctx, req)
+	} else {
+		resp, err = c.resouceManagerClient().ModifyResourceGroup(ctx, req)
+	}
+	if err != nil {
+		return str, err
+	}
+	resErr := resp.GetError()
+	if resErr != nil {
+		return str, errors.Errorf("[pd]" + resErr.Message)
+	}
+	str = resp.GetBody()
+	return
+}
+
+func (c *client) DeleteResourceGroup(ctx context.Context, resourceGroupName string) (string, error) {
+	req := &rmpb.DeleteResourceGroupRequest{
+		ResourceGroupName: resourceGroupName,
+	}
+	resp, err := c.resouceManagerClient().DeleteResourceGroup(ctx, req)
+	if err != nil {
+		return "", err
+	}
+	resErr := resp.GetError()
+	if resErr != nil {
+		return "", errors.Errorf("[pd]" + resErr.Message)
+	}
+	return resp.GetBody(), nil
+}
+
+func (c *client) AcquireTokenBuckets(ctx context.Context, request *rmpb.TokenBucketsRequest) ([]*rmpb.TokenBucketResponse, error) {
+	req := &tokenRequest{
+		done:       make(chan error, 1),
+		requestCtx: ctx,
+		clientCtx:  c.ctx,
+	}
+	req.Requeset = request
+	c.tokenDispatcher.tokenBatchController.tokenRequestCh <- req
+	grantedTokens, err := req.Wait()
+	if err != nil {
+		return nil, err
+	}
+	return grantedTokens, err
+}
+
+type tokenRequest struct {
+	clientCtx    context.Context
+	requestCtx   context.Context
+	done         chan error
+	Requeset     *rmpb.TokenBucketsRequest
+	TokenBuckets []*rmpb.TokenBucketResponse
+}
+
+func (req *tokenRequest) Wait() (tokenBuckets []*rmpb.TokenBucketResponse, err error) {
+	select {
+	case err = <-req.done:
+		err = errors.WithStack(err)
+		if err != nil {
+			return nil, err
+		}
+		tokenBuckets = req.TokenBuckets
+		return
+	case <-req.requestCtx.Done():
+		return nil, errors.WithStack(req.requestCtx.Err())
+	case <-req.clientCtx.Done():
+		return nil, errors.WithStack(req.clientCtx.Err())
+	}
+}
+
+type tokenBatchController struct {
+	tokenRequestCh chan *tokenRequest
+}
+
+func newTokenBatchController(tokenRequestCh chan *tokenRequest) *tokenBatchController {
+	return &tokenBatchController{
+		tokenRequestCh: tokenRequestCh,
+	}
+}
+
+type tokenDispatcher struct {
+	dispatcherCancel     context.CancelFunc
+	tokenBatchController *tokenBatchController
+}
+
+type resourceManagerConnectionContext struct {
+	stream rmpb.ResourceManager_AcquireTokenBucketsClient
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (c *client) createTokenispatcher() {
+	dispatcherCtx, dispatcherCancel := context.WithCancel(c.ctx)
+	dispatcher := &tokenDispatcher{
+		dispatcherCancel: dispatcherCancel,
+		tokenBatchController: newTokenBatchController(
+			make(chan *tokenRequest, 1)),
```

Comment:
What about using a larger size?

Reply:
Currently we have not implemented batch and do not seem to need a larger size